File: python/year2018/day9.py
class Marble:
    def __init__(self, value, prev=None, next=None):
        self.value = value
        self.prev = prev or self
        self.next = next or self

    def insertAfter(self, value):
        marble = Marble(value, self, self.next)
        self.next.prev = marble
        self.next = marble
        return marble

    def delete(self):
        self.prev.next = self.next
        self.next.prev = self.prev
        return self.next

    def rewind(self, n=1):
        i = 0
        marble = self
        while (i < n):
            marble = marble.prev
            i += 1
        return marble

def getWinningScore(players, lastMarble):
    scores = [0] * players
    current = Marble(0)
    i = 1
    while (i < lastMarble):
        if (i % 23 == 0):
            p = (i - 1) % players
            bonus = current.rewind(7)
            scores[p] += i + bonus.value
            current = bonus.delete()
        else:
            current = current.next.insertAfter(i)
        i += 1
    return max(scores)

print getWinningScore(411, 72059 * 100)

File: src/Year2018/Day9.php
<?php

namespace Mzentrale\AdventOfCode\Year2018;

use Mzentrale\AdventOfCode\Puzzle;
use Mzentrale\AdventOfCode\Year2018\Day9\Marble;

class Day9 implements Puzzle
{
    public function part1(string $input)
    {
        list($players, $lastMarble) = $this->parseInput($input);
        return $this->getWinningScoreOptimized($players, $lastMarble);
    }

    public function part2(string $input)
    {
        list($players, $lastMarble) = $this->parseInput($input);
        return $this->getWinningScoreOptimized($players, $lastMarble * 2);
    }

    public function getWinningScore(int $players, int $lastMarble): int
    {
        $scores  = array_fill(0, $players, 0);
        $circle  = [0];
        $current = 0;
        for ($i = 1; $i <= $lastMarble; $i++) {
            $qty = count($circle);
            if ($i % 23 === 0) {
                $p          = ($i - 1) % $players;
                $bonus      = ($current + $qty - 7) % $qty;
                $scores[$p] += $i + $circle[$bonus];
                array_splice($circle, $bonus, 1);
                $current = $bonus;
                continue;
            }

            $current = ($current + 2) % $qty;
            array_splice($circle, $current, 0, $i);
        }
        return max($scores);
    }

    public function getWinningScoreOptimized(int $players, int $lastMarble): int
    {
        $scores  = array_fill(0, $players, 0);
        $current = new Marble(0);
        for ($i = 1; $i <= $lastMarble; $i++) {
            if ($i % 23 === 0) {
                $p          = ($i - 1) % $players;
                $bonus      = $current->prev(7);
                $scores[$p] += $i + $bonus->getValue();
                $current    = $bonus->delete();
                continue;
            }

            $current = $current->next()->insertAfter($i);
        }
        return max($scores);
    }

    private function parseInput(string $input): array
    {
        return (array) sscanf(trim($input), '%d players; last marble is worth %d points');
    }
}
